### PR TITLE
[Parallel intervals] New interval processing

### LINKED ISSUE #226: Support interval syntax with Axes slicing
We would like to support the additional syntax, for example `interval(K[1:])` as an alternative to `interval(1, None)`.

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Replaces the logic in `gt4py.ir.utils.make_axis_interval` and `gt4py.frontend.gtscript_frontend.IRMaker. _visit_interval_node` to support a new syntax and parallel interval parsing.

First PR in the parallel intervals series.

Changes:
- Adds `__getitem__` to `gtscript.Axis` objects so that splitters can be defined, e.g. `K[1]`, and intervals, e.g. `interval(K[1:-1])` (this is now supported!)
- Adds "unit tests" for the interval calls.
- A few unrelated linting fixes.

To do:
- [x] Finish removing old interval processing

Blocked by #244.
Resolves #226.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/gt4py/frontend/gtscript_frontend.py [hunk 1]
```diff
@@ -423,7 +425,7 @@ def visit_Call(self, node: ast.Call, *, target_node=None):
     def visit_Expr(self, node: ast.Expr):
         """ignore pure string statements in callee"""
         if not isinstance(node.value, ast.Str):
-            return super().visit_Expr(node)
+            return super().visit(node)
```
Comment (jdahm) [inline src/gt4py/frontend/gtscript_frontend.py:428]: Unrelated fix that my editor's linter complained about.

File: src/gt4py/ir/nodes.py [hunk 2]
```diff
@@ -685,8 +685,6 @@ class AxisBound(Node):
     level = attribute(of=UnionOf[LevelMarker, VarRef])
     offset = attribute(of=int, default=0)
     loc = attribute(of=Location, optional=True)
-    name: str
-    index: List
```
Comment (jdahm) [inline src/gt4py/ir/nodes.py:689]: Unrelated fix: I couldn't find where these ever get set. They are referenced when using a `VarRef` for the vertical interval, which isn't even fully supported yet, so I am taking this out. `mypy` complains if these lines are removed, but the usage is not, so that is also removed.

File: src/gt4py/gtscript.py [hunk 3]
```diff
@@ -450,7 +487,7 @@ def computation(order):
     return _ComputationContextManager()
 
 
-def interval(start, end):
+def interval(*args):
```
Comment (jdahm) [inline src/gt4py/gtscript.py:490]: Unrelated fix: linter complains on `interval(...)` that there is not value for `end` without this change.

File: src/gt4py/ir/utils.py [hunk 4]
```diff
@@ -30,6 +31,135 @@
 DEFAULT_LAYOUT_ID = "_default_layout_id_"
 
 
+class AxisBoundParser(ast.NodeVisitor):
```
Comment (jdahm) [inline src/gt4py/ir/utils.py:34]: This should really be in `gtscript_frontend.py` but because of the unused `AST2IRVisitor` we need it here to avoid a circular dependency.

@egparedes Can I just remove the unused visitor and disabled tests, or is there logic that we haven't carried over yet?

File: src/gt4py/backend/gt_backends.py [hunk 5]
```diff
@@ -368,9 +368,7 @@ def visit_AxisBound(self, node: gt_ir.AxisBound) -> Tuple[int, int]:
         elif node.level == gt_ir.LevelMarker.END:
             level = len(self.k_splitters) + 1
         else:
-            assert isinstance(node.level, gt_ir.VarRef)
-            assert len(node.level.index) == 1
-            level = self.k_splitters.index((node.name, node.index[0]))
```
Comment (jdahm) [inline src/gt4py/backend/gt_backends.py:373]: This is what had to get removed to satisfy mypy.

**Comment by jdahm** [comment]
@GridTools/gt4py-core This one is ready for review now.

File: src/gt4py/frontend/gtscript_frontend.py (hunk 1 above)
Comment (egparedes) [reply src/gt4py/frontend/gtscript_frontend.py:428]: Yes, the proper way to visit a node is calling the general visit method dispatcher

File: src/gt4py/ir/nodes.py (hunk 2 above)
Comment (egparedes) [reply src/gt4py/ir/nodes.py:689]: Yes, it makes sense since this was never fully implemented. Make sure to update the module docstring with these changes in the `AxisBound` node

File: src/gt4py/backend/gt_backends.py (hunk 5 above)
Comment (egparedes) [reply src/gt4py/backend/gt_backends.py:373]: :+1:

File: src/gt4py/gtscript.py (hunk 3 above)
Comment (egparedes) [reply src/gt4py/gtscript.py:490]: Just a side note here: it the linting problem here is mypy, this could be solved using `overload`
https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading
https://docs.python.org/3/library/typing.html#typing.overload

Comment (jdahm) [reply src/gt4py/gtscript.py:490]: Oh, I like the `@overload` more. I didn't know that existed! I'll make that change.

File: src/gt4py/gtscript.py [hunk 6]
```diff
@@ -329,6 +329,35 @@ def _decorator(func):
...
+    def __repr__(self):
+        return f"_AxisSplitter(axis={self.axis}, offset={self.offset})"
+
+    def __str__(self):
+        return f"{self.axis}[{self.offset}]"
+
+
+class _AxisInterval:
+    def __init__(self, axis: str, start: int, end: int):
+        assert start < end
```
Comment (egparedes) [inline src/gt4py/gtscript.py:346]: Wouldn't this assert break when using negative numbers indexing from then end of the interval like [1:-1]?

File: src/gt4py/frontend/gtscript_frontend.py [hunk 7]
```diff
@@ -655,41 +657,35 @@ def _visit_iteration_order_node(self, node: ast.withitem, loc: gt_ir.Location):
...
-        else:
-            range_node = [interval_node.keywords[0].value, interval_node.keywords[1].value]
-        if len(range_node) == 1 and isinstance(range_node[0], ast.Ellipsis):
-            interval = gt_ir.AxisInterval.full_interval()
-        elif len(range_node) == 2 and all(
-            isinstance(elem, (ast.Num, ast.UnaryOp, ast.NameConstant)) for elem in range_node
+            raise range_error
+        elif (
+            interval.start.level == interval.end.level
+            and interval.end.offset <= interval.start.offset
```
Comment (egparedes) [inline src/gt4py/frontend/gtscript_frontend.py:687]: I think it would be cleaner to  merge these two nested ifs in a single one using `or`

File: tests/test_unittest/test_gtscript_frontend.py [hunk 8]
```diff
@@ -231,6 +232,129 @@ def definition_func(inout_field: gtscript.Field[float]):
...
+            assert def_ir.computations[0].interval.start == gt_ir.AxisBound(
+                level=gt_ir.LevelMarker.START, offset=3, loc=loc
+            )
+            assert def_ir.computations[0].interval.end == gt_ir.AxisBound(
+                level=gt_ir.LevelMarker.END, offset=-1, loc=loc
+            )
+
+    def test_splitter(self):
+        def definition_func(field: gtscript.Field[float]):
+            from __gtscript__ import K
```
Comment (egparedes) [inline tests/test_unittest/test_gtscript_frontend.py:295]: Question: explicit importing K is still not needed, right?

**Review by egparedes — COMMENTED** [review]
It looks good. There are only a couple of questions I would like to understand and the minor issue about updating the ir module docstrings.

File: src/gt4py/gtscript.py (hunk 6 above)
Comment (jdahm) [reply src/gt4py/gtscript.py:346]: Yes! Good catch.

File: src/gt4py/ir/utils.py (hunk 4 above)
Comment (egparedes) [reply src/gt4py/ir/utils.py]: I had missed this comment.... IIRC, the mentioned `AST2IRVisitor` and most of the tools in `ir/utils.py` were just used as helpers to create IR data for some code generation tests. If those tests are not longer used, and since those helpers are not properly documented and unlikely to be needed again, I think it makes sense to remove them, together with other commented out parts in the frontend/ir modules.

If the pieces of dead code to be removed are too large to keep the focus of this PR, another possibility could be to open a new small PR just removing dead code, and then finish this one on top of it.

File: src/gt4py/gtscript.py (hunk 6 above)
Comment (egparedes) [reply src/gt4py/gtscript.py:346]: If this is the case, one of the tests was wrong or incomplete because it was explicitly testing an `[1:-1]` interval

File: tests/test_unittest/test_gtscript_frontend.py (hunk 8 above)
Comment (egparedes) [reply tests/test_unittest/test_gtscript_frontend.py:295]: Actually, another question here: what does `K[2]` mean? It looks like `interval(K[2], -1)` means the same as `interval(2, -1)`, which I find confusing because it is a mix of the old "function-call" notation and the new slice notation.

I also think using the term splitter here is wrong since the resulting AxisBound is still an offset related to the implicit Axis.START level. Something similar in spirit to GridTools splitters would allow to write the AxisBound as `AxisBound(level=splitter, offset=0)` and then the syntax could be either `interval(splitter, -1)` or `interval(K[splitter:-1])` being splitter an int variable (external or local, or even run-time parameter if we allow this in the future).

File: src/gt4py/ir/utils.py (hunk 4 above)
Comment (jdahm) [reply src/gt4py/ir/utils.py]: I like that approach. I'll do that today.

Comment (jdahm) [reply src/gt4py/ir/utils.py]: See #244.

File: tests/test_unittest/test_gtscript_frontend.py (hunk 8 above)
Comment (jdahm) [reply tests/test_unittest/test_gtscript_frontend.py:295]: I can try to make a change that requires EITHER the old or new syntax to be used, but not either. Would that be a good approach?

File: src/gt4py/gtscript.py (hunk 3 above)
Comment (jdahm) [reply src/gt4py/gtscript.py:490]: I can't seem to get this to work. I keep getting the error:

```
error: An overloaded function outside a stub file must have an implementation
```

File: src/gt4py/gtscript.py (hunk 6 above)
Comment (jdahm) [reply src/gt4py/gtscript.py:346]: Ah, it never actually makes the AxisInterval in that case -- it only parses the AST as if it were one. If the `K[1:-1]` were a named external, it would have caught the code error.

File: tests/test_unittest/test_gtscript_frontend.py (hunk 8 above)
Comment (jdahm) [reply tests/test_unittest/test_gtscript_frontend.py:295]: > I also think using the term splitter here is wrong since the resulting AxisBound is still an offset related to the implicit Axis.START level.

Good point. I'll rename the object that e.g. `gtscript.K[2]` returns to be a `gtscript._AxisOffset`.

Comment (jdahm) [reply tests/test_unittest/test_gtscript_frontend.py:295]: > what does K[2] mean? It looks like interval(K[2], -1) means the same as interval(2, -1), which I find confusing because it is a mix of the old "function-call" notation and the new slice notation.

I agree that mixing up the old and new style is confusing here. The technical detail is that if the `K[2]` were a an external instead of inlined (a `gtscript._AxisOffset` object), then the `ValueInliner` replaces this by an `ast.Constant` node before the `IRMaker` runs (which calls the `AxisIntervalParser`). That is to say that in the case of an external we cannot distinguish between the new and old style syntax.

I can add a check that in the case of old-style syntax AND not using named offsets (externals), then enforce that new and old style syntax is not mixed. I'll make that change now.

File: src/gt4py/ir/nodes.py (hunk 2 above)
Comment (jdahm) [reply src/gt4py/ir/nodes.py:689]: There was no reference in the module docstring.

**Review by egparedes — APPROVED** [review]
It looks ready to me. Do you want to merge the new master here before merging the PR?
